File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    # Type: “world” or "room"
    channel_type = params[:channel_type] || "world"
    room_id = params[:room_id] if channel_type == "room" # Chat with players in the same grid

    if channel_type == "world"
      stream_from "world_chat"
      Rails.logger.info "Subscribed to world_chat"
    elsif channel_type == "room" && room_id.present?
      stream_from "room_chat_#{room_id}"
      Rails.logger.info "Subscribed to room_chat_#{room_id}"
    else
      reject
    end
  end

  def receive(data)
    if data["message"].present?
      # print, debug, log
      Rails.logger.info "Message received: #{data['message']}"
      channel_type = params[:channel_type] || "world"
      room_id = params[:room_id]
      username = current_user.character.character_name

      # Save message into database
      Message.create!(
        content: data["message"],
        username: username,
        channel_type: channel_type,
        room_id: room_id
      )

      if channel_type == "world"
        # Broadcast to the world chat
        ActionCable.server.broadcast("world_chat", { username: username, message: data["message"], channel: "world" })
      elsif channel_type == "room" && room_id.present?
        # Broadcast to room chat
        ActionCable.server.broadcast("room_chat_#{room_id}", { username: username, message: data["message"], channel: "room" })
      end
    end
  end

  def load_recent_messages(data)
    maximum_message = 50
    channel_type = params[:channel_type]
    room_id = params[:room_id]

    messages = if channel_type == "world"
                 Message.where(channel_type: "world").order(created_at: :desc).limit(maximum_message)
               elsif channel_type == "room" && room_id.present?
                 Message.where(channel_type: "room", room_id: room_id).order(created_at: :desc).limit(maximum_message)
               else
                 []
               end

    transmit(messages.reverse.map { |msg| { username: msg.username, message: msg.content } })
  end
end

File: app/javascript/channels/chat_channel.js
import consumer from "channels/consumer";

let chatChannel = null;

function subscribeToChannel(channelType, roomName = null) {
  if (chatChannel) {
    chatChannel.unsubscribe();
  }

  chatChannel = consumer.subscriptions.create(
      {channel: "ChatChannel", channel_type: channelType, room_id: roomName},
      {
        received(data) {
          const messagesContainer = document.getElementById("messages");
          if (messagesContainer) {
            if (Array.isArray(data)) {
              // Historical message
              clearMessages();
              data.forEach((msg) => renderMessage(msg, messagesContainer));
            } else {
              // Real-time message
              renderMessage(data, messagesContainer);
            }
            // Automatically scroll to the latest messages
            messagesContainer.scrollTop = messagesContainer.scrollHeight;
          }
        },
        connected() {
          // Load recent messages
          console.log(`Connected to ${channelType} channel`);
          clearMessages();
          chatChannel.perform("load_recent_messages", {}); // Load recent message
        },
        disconnected() {
          console.log("Disconnected from chat channel.");
        },
      }
  );
}

function clearMessages() {
  const messagesContainer = document.getElementById("messages");
  if (messagesContainer) {
    messagesContainer.innerHTML = "";
  }
}

function renderMessage(data, messagesContainer){
  const messageElement = document.createElement("div");
  const usernameElement = document.createElement("span");
  usernameElement.innerText = `${data.username}: `;
  usernameElement.classList.add("username");

  const messageContent = document.createElement("span");
  messageContent.innerText = data.message;

  messageElement.appendChild(usernameElement);
  messageElement.appendChild(messageContent);
  messagesContainer.appendChild(messageElement);
}

document.addEventListener("turbo:load", () => {
  if (!document.body.classList.contains("chat-page")) {
    return;
  }

  const chatForm = document.getElementById("chat-form");
  const chatInput = document.getElementById("chat-input");
  const chatTypeSelector = document.getElementById("chat-type-selector");

  // Initialize as world chat
  subscribeToChannel("world");

  // Change the chat type
  if (chatTypeSelector) {
    chatTypeSelector.addEventListener("change", (event) => {
      const selectedType = event.target.value; // Get chat type that user select
      if (selectedType === "world") {
        subscribeToChannel("world");
      } else if (selectedType === "room") {
        const roomId = document.body.dataset.roomId; // Get roomID
        subscribeToChannel("room", roomId);
      }
    });
  }

  // Send message
  if (chatForm && chatInput) {
    chatForm.addEventListener("submit", (event) => {
      event.preventDefault();
      const message = chatInput.value.trim();
      if (message !== "" && chatChannel) {
        console.log("Sending message:", message);
        chatChannel.send({ message });
        chatInput.value = "";
      }
    });
  } else {
    console.warn("Chat form or input not found.");
  }
});

